### PR TITLE
FIX Cast absoluteUrl() argument to string

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripeltd/product-issues/issues/642

Similar to https://github.com/silverstripe/silverstripe-elemental/pull/1029

### CODE REVIEW

File: tests/php/Control/DirectorTest.php [hunk 1]
```diff
@@ -97,9 +97,9 @@ public function testAbsoluteURL()
 
         //test empty / local urls
         foreach (['', './', '.'] as $url) {
-            $this->assertEquals("http://www.mysite.com:9090/mysite/", Director::absoluteURL($url, Director::BASE));
-            $this->assertEquals("http://www.mysite.com:9090/", Director::absoluteURL($url, Director::ROOT));
-            $this->assertEquals("http://www.mysite.com:9090/mysite/sub-page/", Director::absoluteURL($url, Director::REQUEST));
+            $this->assertEquals("http://www.mysite.com:9090/mysite/", Director::absoluteURL((string) $url, Director::BASE));
+            $this->assertEquals("http://www.mysite.com:9090/", Director::absoluteURL((string) $url, Director::ROOT));
+            $this->assertEquals("http://www.mysite.com:9090/mysite/sub-page/", Director::absoluteURL((string) $url, Director::REQUEST));
```
Comment (GuySartorelli) [inline tests/php/Control/DirectorTest.php]:
```suggestion
            $this->assertEquals("http://www.mysite.com:9090/mysite/", Director::absoluteURL($url, Director::BASE));
            $this->assertEquals("http://www.mysite.com:9090/", Director::absoluteURL($url, Director::ROOT));
            $this->assertEquals("http://www.mysite.com:9090/mysite/sub-page/", Director::absoluteURL($url, Director::REQUEST));
```
Not necessary here.

File: tests/php/Control/ControllerTest.php [hunk 2]
```diff
@@ -475,7 +475,7 @@ public function testRedirectBackByBackUrl()
         $externalAbsoluteUrl = 'http://myhost.com/some-url';
         $response = $this->get('TestController/redirectbacktest?BackURL=' . urlencode($externalAbsoluteUrl ?? ''));
         $this->assertEquals(
-            Director::absoluteURL($link),
+            Director::absoluteURL((string) $link),
```
Comment (GuySartorelli) [inline tests/php/Control/ControllerTest.php]:
```suggestion
            Director::absoluteURL($link),
```
Not necessary here

File: src/Security/PasswordExpirationMiddleware.php [hunk 3]
```diff
@@ -145,15 +145,15 @@ protected function handleExpiredPassword(HTTPRequest $request): ?HTTPResponse
             return null;
         }
 
-        $currentUrl = $this->absoluteUrl($request->getURL(true));
+        $currentUrl = $this->absoluteUrl((string) $request->getURL(true));
```
Comment (GuySartorelli) [inline src/Security/PasswordExpirationMiddleware.php]:
```suggestion
        $currentUrl = $this->absoluteUrl($request->getURL(true));
```
Not necessary here

File: src/Forms/HTMLEditor/TinyMCEConfig.php [hunk 4]
```diff
@@ -863,7 +863,7 @@ public static function get_tinymce_lang_url(): string
         $dir = static::config()->get('lang_dir');
         if ($lang !== 'en' && !empty($dir)) {
             $resource = ModuleResourceLoader::singleton()->resolveResource($dir);
-            return Director::absoluteURL($resource->getRelativeResource($lang . '.js')->getURL());
+            return Director::absoluteURL((string) $resource->getRelativeResource($lang . '.js')->getURL());
```
Comment (GuySartorelli) [inline src/Forms/HTMLEditor/TinyMCEConfig.php]:
```suggestion
            return Director::absoluteURL($resource->getRelativeResource($lang . '.js')->getURL());
```
Not necessary here

File: src/Forms/HTMLEditor/TinyMCEConfig.php [hunk 5]
```diff
@@ -675,7 +675,7 @@ protected function getConfig()
                 );
             } elseif (!Director::is_absolute_url($path)) {
                 // Non-absolute urls are made absolute
-                $path = Director::absoluteURL($path);
+                $path = Director::absoluteURL((string) $path);
```
Comment (GuySartorelli) [inline src/Forms/HTMLEditor/TinyMCEConfig.php:678]:
```suggestion
                $path = Director::absoluteURL($path);
```
Not needed here

File: src/Forms/HTMLEditor/TinyMCEConfig.php [hunk 6]
```diff
@@ -666,7 +666,7 @@ protected function getConfig()
         $plugins = [];
         foreach ($this->getPlugins() as $plugin => $path) {
             if ($path instanceof ModuleResource) {
-                $path = Director::absoluteURL($path->getURL());
+                $path = Director::absoluteURL((string) $path->getURL());
```
Comment (GuySartorelli) [inline src/Forms/HTMLEditor/TinyMCEConfig.php]:
```suggestion
                $path = Director::absoluteURL($path->getURL());
```
Not needed here

File: src/Control/RequestHandler.php [hunk 7]
```diff
@@ -581,7 +581,7 @@ public function Link($action = null)
      */
     public function redirect(string $url, int $code = 302): HTTPResponse
     {
-        $url = Director::absoluteURL($url);
+        $url = Director::absoluteURL((string) $url);
```
Comment (GuySartorelli) [inline src/Control/RequestHandler.php]:
```suggestion
        $url = Director::absoluteURL($url);
```
Not needed here

File: src/Control/HTTP.php [hunk 8]
```diff
@@ -88,7 +88,7 @@ public static function absoluteURLs($html)
      * for example:
      * <code>
      * function($url) {
-     *      return Director::absoluteURL($url, true);
+     *      return Director::absoluteURL((string) $url, true);
```
Comment (GuySartorelli) [inline src/Control/HTTP.php]: Lets do `function(string $url)` here instead to promote good standards

File: src/Control/HTTP.php [hunk 9]
```diff
@@ -52,7 +52,7 @@ public static function filename2url($filename)
         }
 
         $relativePath = ltrim(substr($filename ?? '', $baseLength ?? 0), '/\\');
-        return Director::absoluteURL($relativePath);
+        return Director::absoluteURL((string) $relativePath);
```
Comment (GuySartorelli) [inline src/Control/HTTP.php]:
```suggestion
        return Director::absoluteURL($relativePath);
```

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
See https://github.com/silverstripe/silverstripe-cms/pull/2806#discussion_r1046613902 for a discussion on this approach. Strongly typing the return type of the given methods is out of scope for CMS 5.

Some to remove here where it's not necessary though. It's a fairly small PR so I think it's reasonable to only do this where the variable can actually be something other than a string.

File: src/Control/HTTP.php (hunk 9 above)
Comment (emteknetnz) [reply src/Control/HTTP.php]: Updated

File: src/Control/HTTP.php (hunk 8 above)
Comment (emteknetnz) [reply src/Control/HTTP.php]: Updated

File: src/Control/RequestHandler.php (hunk 7 above)
Comment (emteknetnz) [reply src/Control/RequestHandler.php]: Updated

File: src/Forms/HTMLEditor/TinyMCEConfig.php (hunk 6 above)
Comment (emteknetnz) [reply src/Forms/HTMLEditor/TinyMCEConfig.php]: Updated

File: src/Forms/HTMLEditor/TinyMCEConfig.php (hunk 5 above)
Comment (emteknetnz) [reply src/Forms/HTMLEditor/TinyMCEConfig.php:678]: I think this one does - $this->plugins has a bunch of `null` $paths

Comment (GuySartorelli) [reply src/Forms/HTMLEditor/TinyMCEConfig.php:678]: Ahh yup, sorry, I misread this one as coming from the `join_link` above which it definitely doesn't. Leave this one in

File: src/Forms/HTMLEditor/TinyMCEConfig.php (hunk 4 above)
Comment (emteknetnz) [reply src/Forms/HTMLEditor/TinyMCEConfig.php]: Updated

File: src/Security/PasswordExpirationMiddleware.php (hunk 3 above)
Comment (emteknetnz) [reply src/Security/PasswordExpirationMiddleware.php]: Updated

File: tests/php/Control/ControllerTest.php (hunk 2 above)
Comment (emteknetnz) [reply tests/php/Control/ControllerTest.php]: Updated

File: tests/php/Control/DirectorTest.php (hunk 1 above)
Comment (emteknetnz) [reply tests/php/Control/DirectorTest.php]: Updated

File: src/Security/PasswordExpirationMiddleware.php (hunk 3 above)
Comment (emteknetnz) [reply src/Security/PasswordExpirationMiddleware.php]: Actually adding it back it - it caused a unit test to fail - https://github.com/silverstripe/silverstripe-framework/actions/runs/3689681730/jobs/6245906511#step:12:116

Comment (GuySartorelli) [reply src/Security/PasswordExpirationMiddleware.php]: I'd be keen to understand how this cast-to-string is affecting that unit test - I would have expected the CI to fail by throwing an exception if the problem was that the wrong type is being passed to a strongly-typed parameter.

**Review by GuySartorelli — APPROVED** [review]

**Pull request merged**